primitives/arithmetic/src/fixed128.rs

```diff
@@ -0,0 +1,351 @@
+	/// Consume self and return the inner value.
+	///
+	/// This should only be used for testing.
+	#[cfg(any(feature = "std", test))]
+	pub fn into_inner(self) -> i128 { self.0 }
+
+	/// Raw constructor. Equal to `parts / 1_000_000_000_000_000_000`.
+	pub fn from_parts(parts: i128) -> Self {
+		Self(parts)
+	}
+
+	/// creates self from a rational number. Equal to `n/d`.
+	///
+	/// Note that this might be lossy.
+	pub fn from_rational(n: i128, d: u128) -> Self {
+		// Turn n into a positive number, and track the original sign.
+		// This lossy by 1 if n is exactly i128.min_value()
+		let (n, sign): (u128, i128) = if n == i128::min_value() {
+			(i128::max_value().try_into().expect("i128 max value can fit into u128"), -1)
+		} else if n.is_negative() {
+			((n * -1).try_into().expect("positive i128 can always fit into u128"), -1)
+		} else {
+			(n.try_into().expect("positive i128 can always fit into u128"), 1)
+		};
+		// This should never fail, but if it does, it is because value was too big for u128
+		// so we map to u128 max value. TODO: Can we expect here that value is always smaller?
+		let result = multiply_by_rational(n, U_DIV, d.max(1)).unwrap_or(u128::max_value());
+		// If the u128 result wont fit in i128, we do i128 max. TODO: Can we expect value is always smaller?
+		let signed_result: i128 = sign * result.try_into().unwrap_or(i128::max_value());
+		Self(signed_result)
+	}
+
+	/// Performs a saturated multiply and accumulate by unsigned number.
+	///
+	/// Returns a saturated `int + (self * int)`.
+	pub fn saturated_multiply_accumulate<N>(self, int: N) -> N
+		where
+			N: TryFrom<u128> + From<u64> + UniqueSaturatedInto<u64> + Bounded + Clone + Saturating +
+			ops::Rem<N, Output=N> + ops::Div<N, Output=N> + ops::Mul<N, Output=N> +
+			ops::Add<N, Output=N>,
+	{
+		let div = DIV as u128;
+		let positive = self.0 > 0;
+		// safe to convert as absolute value.
+		let parts = self.0.checked_abs().map(|v| v as u128).unwrap_or(i128::max_value() as u128 + 1);
+
+
+		// will always fit.
+		let natural_parts = parts / div;
+		// might saturate.
+		let natural_parts: N = natural_parts.saturated_into();
+		// fractional parts can always fit into u64.
+		let perquintill_parts = (parts % div) as u64;
+
+		let n = int.clone().saturating_mul(natural_parts);
+		let p = Perquintill::from_parts(perquintill_parts) * int.clone();
+
+		// everything that needs to be either added or subtracted from the original weight.
+		let excess = n.saturating_add(p);
+
+		if positive {
+			int.saturating_add(excess)
+		} else {
+			int.saturating_sub(excess)
+		}
+	}
+}
+
+impl Saturating for Fixed128 {
+	fn saturating_add(self, rhs: Self) -> Self {
+		Self(self.0.saturating_add(rhs.0))
+	}
+
+	fn saturating_mul(self, rhs: Self) -> Self {
+		Self(self.0.saturating_mul(rhs.0) / DIV)
+	}
+
+	fn saturating_sub(self, rhs: Self) -> Self {
+		Self(self.0.saturating_sub(rhs.0))
+	}
+
+	fn saturating_pow(self, exp: usize) -> Self {
+		Self(self.0.saturating_pow(exp as u32))
```

Comment:
This is one place where we did not change `u32` to `u64` since `usize` cannot cleanly convert.